service/dap/server.go

```diff
@@ -3042,10 +3042,76 @@ func (s *Session) onTerminateRequest(request *dap.TerminateRequest) {
 	s.sendNotYetImplementedErrorResponse(request.Request)
 }
 
-// onRestartRequest sends a not-yet-implemented error response
-// Capability 'supportsRestartRequest' is not set in 'initialize' response.
+// onRestartRequest handles 'restart' request.
+// Capability 'supportsRestartRequest' is set to true in 'initialize' response.
 func (s *Session) onRestartRequest(request *dap.RestartRequest) {
-	s.sendNotYetImplementedErrorResponse(request.Request)
+	s.changeStateMu.Lock()
+	defer s.changeStateMu.Unlock()
+
+	// Cannot restart in noDebug mode
+	if s.isNoDebug() {
+		s.sendErrorResponse(request.Request, FailedToLaunch, "Cannot restart", "cannot restart in noDebug mode")
+		return
+	}
+
+	if s.debugger == nil {
+		s.sendErrorResponse(request.Request, FailedToLaunch, "Cannot restart: debugger is not running", "Debugger is not running")
+		return
+	}
+
+	// Cannot restart attached processes
+	if s.debugger.AttachPid() != 0 {
+		s.sendErrorResponse(request.Request, FailedToLaunch, "Cannot restart", "cannot restart process Delve did not create")
+		return
+	}
+
+	var newArgs []string
+	var resetArgs bool
+
+	// Update launch/attach arguments if provided
+	if len(request.Arguments) > 0 {
+		// RestartArguments has an "arguments" field that contains launch/attach config
+		var restartArgs struct {
```

Comment:
Can we do this whole thing in a single unmarshal, by unmarshalling into something like this:

```
type restartArgs struct {
    Arguments struct {
        Request string `json:"request"`
        LaunchConfig
    }
}
```

also if the unmarshal fails shouldn't we return an error instead of ignoring the arguments and doing a restart?